src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1743,6 +1758,35 @@ string dumpDagTest(BoundDagTest d)
 #endif
         }
 
+        /// <summary>
+        /// This is a readonly wrapper around an array builder.  It ensures we can benefit from the pooling an array builder providers, without having to incur 
+        /// intermediary allocations for <see cref="ImmutableArray"/>s.
+        /// </summary>
+        private readonly struct FrozenArrayBuilder<T>
+        {
+            private readonly ArrayBuilder<T> _arrayBuilder;
+
+            public FrozenArrayBuilder(ArrayBuilder<T> arrayBuilder)
```

Comment:
No need to make a change here. But what springs to my mind is `ArrayBuilder<T>.AsFrozen()` for general usage in the future.

I feel good about us starting to experiment with stuff like this by just introducing it in this limited scope.

Reply:
Agreed completely